src/xrpld/app/paths/detail/MPTEndpointStep.cpp

```diff
@@ -0,0 +1,1056 @@
+{
+    // Unlike IOU, MPT can't exceed the limit (MaximumAmount).
+    // See DirectIOfferCrossingStep::maxFlow().
+    return maxPaymentFlow(sb);
+}
+
+TER
+MPTEndpointPaymentStep::check(
+    StrandContext const& ctx,
+    std::shared_ptr<const SLE> const& sleSrc) const
+{
+    // Since this is a payment, MPToken must be present.  Perform all
+    // MPToken related checks.
+
+    // requireAuth checks if MPTIssuance exist. Note that issuer to issuer
+    // payment is invalid
+    auto const& issuer = mptIssue_.getIssuer();
+    if (src_ != issuer)
+    {
+        if (auto const ter = requireAuth(ctx.view, mptIssue_, src_);
+            ter != tesSUCCESS)
+            return ter;
+    }
+
+    if (dst_ != issuer)
+    {
+        if (auto const ter = requireAuth(ctx.view, mptIssue_, dst_);
+            ter != tesSUCCESS)
+            return ter;
+    }
+
+    // Direct MPT payment, no DEX
+    if (mptIssue_ == ctx.strandDeliver &&
+        (ctx.isFirst || (ctx.prevStep && !ctx.prevStep->bookStepBook())))
+    {
+        // Between holders
+        if (isDirectBetweenHolders_)
+        {
+            auto const& holder = ctx.isFirst ? src_ : dst_;
+            // Payment between the holders
+            if (isFrozen(ctx.view, holder, mptIssue_))
+                return tecLOCKED;
+
+            if (auto const ter =
+                    canTransfer(ctx.view, mptIssue_, holder, ctx.strandDst);
+                ter != tesSUCCESS)
+                return ter;
+        }
+        // Don't need to check if a payment is between issuer and holder
+        // in either direction
+    }
+    // Cross-token MPT payment via DEX
+    else
+    {
+        auto const account = ctx.isFirst ? src_ : dst_;
+        if (auto const ter = isMPTDEXAllowed(
+                ctx.view, mptIssue_, account, mptIssue_.getIssuer());
+            ter != tesSUCCESS)
+            return ter;
+    }
+
+    return tesSUCCESS;
+}
+
+TER
+MPTEndpointOfferCrossingStep::check(
+    StrandContext const& ctx,
+    std::shared_ptr<const SLE> const&) const
+{
+    auto const& holder = ctx.isFirst ? src_ : dst_;
+    auto const& issuer = mptIssue_.getIssuer();
+    if (holder != issuer)
+    {
+        if (auto const ter =
+                isMPTDEXAllowed(ctx.view, mptIssue_, holder, issuer);
+            ter != tesSUCCESS)
+            return ter;
+    }
+    return tesSUCCESS;
+}
+
+TER
+MPTEndpointOfferCrossingStep::checkCreateMPT(
+    ApplyView& view,
+    ripple::DebtDirection srcDebtDir)
+{
+    // TakerPays is the last step if offer crossing
+    if (isLast_)
+    {
+        // Create MPToken for the offer's owner. No need to check
+        // for the reserve since the offer doesn't go on the books
+        // if crossed. Insufficient reserve is allowed if the offer
+        // crossed. See CreateOffer::applyGuts() for reserve check.
+        if (auto const err =
+                MPTokenAuthorize::checkCreateMPT(view, mptIssue_, dst_, j_);
+            err != tesSUCCESS)
```

Comment:
This does not check if MPTIssuance enabled authorization. I think we should only allow creation of MPToken if MPTIssuance has **not** set `lsfMPTRequireAuth`

Reply:
`CreateOffer` checks for auth in `CreateOffer::checkAcceptAsset`